avocado/core/nrunner.py

```diff
@@ -613,97 +569,166 @@ def wait(self):
-    task = Task.from_recipe(args.get('recipe'))
-    task_run(task, echo)
+    CMD_TASK_RUN_ARGS = (
+        (("-i", "--identifier"),
+         {'type': str, 'required': True, 'help': 'Task unique identifier'}),
+        (("-s", "--status-uri"),
+         {'action': "append", 'default': None,
+          'help': 'URIs of status services to report to'}),
+    )
+    CMD_TASK_RUN_ARGS += CMD_RUNNABLE_RUN_ARGS
 
+    CMD_TASK_RUN_RECIPE_ARGS = (
+        (("recipe", ),
+         {'type': str, 'help': 'Path to the task recipe file'}),
+    )
 
-CMD_STATUS_SERVER_ARGS = (
-    (("uri", ),
-     {'type': str, 'help': 'URI to bind a status server to'}),
+    CMD_STATUS_SERVER_ARGS = (
+        (("uri", ),
+         {'type': str, 'help': 'URI to bind a status server to'}),
     )
 
+    def __init__(self, echo=print, prog=None, description=None):
+        self.echo = echo
+        self.parser = None
+        if prog is None:
+            prog = self.PROG_NAME
+        if description is None:
+            description = self.PROG_DESCRIPTION
+        self._setup_parser(prog, description)
+
+    def _setup_parser(self, prog, description):
+        self.parser = argparse.ArgumentParser(prog=prog,
+                                              description=description)
+        subcommands = self.parser.add_subparsers(dest='subcommand')
+        subcommands.required = True
+        for cmd_meth in self._get_commands_method_without_prefix():
+            attr = "CMD_%s_ARGS" % cmd_meth.upper()
+            cmd = cmd_meth.replace('_', '-')
+            cmd_parser = subcommands.add_parser(cmd)
+            if hasattr(self, attr):
+                for arg in getattr(self, attr):
+                    cmd_parser.add_argument(*arg[0], **arg[1])
+
+    def _get_commands_method_without_prefix(self):
+        prefix = 'command_'
+        return [c[0][len(prefix):]
+                for c in inspect.getmembers(self, inspect.ismethod)
+                if c[0].startswith(prefix)]
+
+    def run(self):
+        args = vars(self.parser.parse_args())
+        subcommand = args.get('subcommand')
+        if subcommand in self.get_commands():
+            meth_name = 'command_' + subcommand.replace('-', '_')
+            if hasattr(self, meth_name):
+                kallable = getattr(self, meth_name)
+                return kallable(args)
+
+    def get_commands(self):
+        return [c.replace('_', '-') for c in
+                self._get_commands_method_without_prefix()]
+
+    def get_capabilities(self):
+        return {"runnables": [k for k in self.RUNNABLE_KINDS_CAPABLE.keys()],
+                "commands": self.get_commands()}
+
+    def get_runner_from_runnable(self, runnable):
+        runner = self.RUNNABLE_KINDS_CAPABLE.get(runnable.kind, None)
+        if runner is not None:
+            return runner(runnable)
+        raise ValueError('Unsupported kind of runnable: %s' % runnable.kind)
+
+    def command_capabilities(self, _):
```

Comment:
What you think of having a better name here? An action? `show_capabilities` for instance? If the idea is not to have as a clear "action/method", then I suggest adding the `@property` decorator. For this and all the methods below.

Also, I understand the `_` for "garbage/temporary" variables. But I could not find any part of the code using this method.

Could you elaborate on why are you using `_` here?

Reply:
WRT naming, I find that the usability of the command line is simpler and better like this:

```
$ avocado-runner capabilities
```

Than with:

```
$ avocado-runner show-capabilities
```

In this specific case, I don't think "show" adds much. I see it similarly to `app --version` instead of `app --show-version`.

About the treating those as properties, I don't understand why I'd do it, given that I want to pass parameters to those `command_` methods. Maybe I'm missing something.

About the garbage variable, the reason is that, for simplification, all command methods share the same signature, and receive the parsed args. The `command_capabilities` method, though, doesn't use it.

Reply:
@clebergnu 

I agree with the command line convention. My suggestion was regarding the method name. Not sure, why the user interface needs to be a 1-1 mapping on the code methods.

I suggested the @property, because `_` was not being used, so maybe we could change this method signature.

Also, today I was asking myself what could be the reason for the "capabilities" command, since that -h will show you all the subcommands, right? Unless that I'm missing something, it looks like redundant.

Again, this is just a suggestion.